app/models/manageiq/providers/openstack/infra_manager/host.rb

```diff
@@ -35,48 +40,84 @@ def get_parent_keypair(type = nil)
   def authentication_best_fit(type = nil)
     # First check for Host level credentials with filled private key, this way we can override auth credentials per
     # host. Otherwise host level auth will hold only state of auth for the given host. We allow only auth with private
-    # key for OpenstackInfra hosts
-    auth = authentication_type(type)
-    return auth if auth && auth.auth_key
+    # key and password for OpenstackInfra hosts
+    auth = if !authentication_type(type).try(:password).blank? || !authentication_type(type).try(:auth_key).blank?
+             authentication_type(type)
+           elsif !authentication_type(:ssh_keypair).try(:auth_key).blank?
+             authentication_type(:ssh_keypair)
+           elsif !authentication_type(:default).try(:password).blank?
+             authentication_type(:default)
+           end
+
+    return auth if auth && (auth.password || auth.auth_key)
```

Comment:
Using a way of @matthewd , @blomquisg your way was not passing tests. :-)

There is a bit more general check for credentials now, but it should be ok.

Thank you guys, it's much nicer now. :-)